File: chapter02/2.14/Scala/Main.scala
/*

Run with ScalaCLI:

    scala-cli Main.scala

[Scala]$ time scala-cli ./Main.scala
Compiling project (Scala 3.5.1, JVM (17))
Compiled project (Scala 3.5.1, JVM (17))
4 Queens: 365596

real    0m27,671s
user    0m28,098s
sys     0m0,455s

 */

def solve(cols: Set[Int], d1: Set[Int], d2: Set[Int]): Int =
  if cols.isEmpty then 1
  else
    cols
      .diff(d1)
      .diff(d2)
      .foldLeft(0)((acc, n) =>
        acc + solve(cols - n, (d1 + n).map(_ + 1), (d2 + n).map(_ - 1))
      )

@main
def main(args: String*): Unit =
  println(
    "Possible solutions for 14 Queens: " + solve(
      Set(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14),
      Set(),
      Set()
    )
  )

File: chapter02/2.14/main.rb
require "immutable/set"

require 'awesome_print'
require 'pry'

def solve(cols, d1, d2)
  if cols.empty?
    1
  else
    cols.difference(d1).difference(d2).reduce(0) do |acc, n|
      acc +
        solve(cols.delete(n),
              d1.add(n).map { |x| x + 1 },
              d2.add(n).map { |x| x - 1 }
             )
    end
  end
end

=begin

The performance is abysimal in comparaison to Haskell or OCaml

[2.14]$ time ruby ./main.rb
365596

real    9m29,389s
user    9m27,560s
sys     0m0,993s

=end
p(solve(Immutable::Set[*(1..14).to_a], Immutable::Set.empty, Immutable::Set.empty))
